src/MetricsUploader/MetricsUploader.cpp

```diff
@@ -6,31 +6,55 @@
 
 #include <vector>
 
+#include "MetricsUploader/Result.h"
+#include "OrbitBase/ImmediateExecutor.h"
 #include "OrbitBase/Logging.h"
 #include "OrbitVersion/OrbitVersion.h"
 
 namespace orbit_metrics_uploader {
 
+using orbit_base::Future;
+
 bool MetricsUploader::SendLogEvent(OrbitLogEvent_LogEventType log_event_type,
-                                   std::chrono::milliseconds event_duration) {
-  if (send_log_event_addr_ != nullptr) {
-    OrbitLogEvent log_event;
-    log_event.set_log_event_type(log_event_type);
-    log_event.set_orbit_version(orbit_core::GetVersion());
-    log_event.set_event_duration_milliseconds(event_duration.count());
-    log_event.set_session_uuid(session_uuid_);
-
-    int message_size = log_event.ByteSize();
-    std::vector<uint8_t> buffer(message_size);
-    log_event.SerializeToArray(buffer.data(), message_size);
-
-    Result result = send_log_event_addr_(buffer.data(), message_size);
-    if (result == kNoError) {
-      return true;
-    }
-    ERROR("Can't start the metrics uploader client: %s", GetErrorMessage(result));
+                                   std::chrono::milliseconds event_duration,
+                                   OrbitLogEvent_StatusCode status_code) {
+  if (send_log_event_addr_ == nullptr) {
+    ERROR("Unable to send metric, send_log_event_addr_ is nullptr");
+    return false;
+  }
+
+  OrbitLogEvent log_event;
+  log_event.set_log_event_type(log_event_type);
+  log_event.set_orbit_version(orbit_core::GetVersion());
+  log_event.set_event_duration_milliseconds(event_duration.count());
+  log_event.set_session_uuid(session_uuid_);
+  log_event.set_status_code(status_code);
+
+  int message_size = log_event.ByteSize();
+  std::vector<uint8_t> buffer(message_size);
+  log_event.SerializeToArray(buffer.data(), message_size);
+
+  Result result = send_log_event_addr_(buffer.data(), message_size);
+  if (result != kNoError) {
+    ERROR("Unable to send metrics event: %s", GetErrorMessage(result));
+    return false;
   }
-  return false;
+
+  return true;
+}
+
+Future<bool> MetricsUploader::SendTimedStatusEvent(OrbitLogEvent_LogEventType log_event_type,
+                                                   const Future<OrbitLogEvent_StatusCode>& future) {
+  auto timestamp_before = std::chrono::steady_clock::now();
+
+  orbit_base::ImmediateExecutor executer{};
+  return future.Then(&executer, [timestamp_before, log_event_type,
+                                 this](OrbitLogEvent_StatusCode status_code) {
+    auto timestamp_after = std::chrono::steady_clock::now();
+    auto duration =
+        std::chrono::duration_cast<std::chrono::milliseconds>(timestamp_after - timestamp_before);
+    return SendLogEvent(log_event_type, duration, status_code);
+  });
```

Comment:
I'm not sure if a future is the best mechanism here. Usually you would use a RAII here. Something like that:

```cpp
// Sends event with duration when it gets destructed
class ScopedMetric {
    std::chrono::steady_clock::timestamp before_;
    MetricsUploader* uploader_ = nullptr;
 public:
    explicit ScopedMetric(MetricsUploader* uploader) : before_{std::chrono::steady_clock::now()}, uploader_{uploader} {}
    ~ScopedMetric() noexcept {
      if (!uploader_) return;
      auto duration = std::chrono::steady_clock::now() - before_;
      uploader_->SendLogEvent(...);
    }
 };
```